merge pbjx-token branch to master

albertgcsula opened the pull request:
- add determineEventSource and getConfig utilities for AWS Lambda use
- add TransportEnvelope js code (php lib counterpart)
- add test and test cases for determineEventSource, getConfig and TransportEnvelop methods

gdbrown committed:
stub aws lambda helpers

gdbrown committed:
stub transport envelope as well

Commit:
move transport envelope to pbjx-token branch

gdbrown committed:
stub tests for aws helpers and transport envelope

Commit:
add getconfig method

base64 committed:
integrate pbjxToken, create determineEventSources and associated unit test

base64 committed:
missing npm packages

base64 committed:
include JWT HMAC SHA256 Token test suite

base64 committed:
eslint fixes

base64 committed:
eslint fixes

base64 committed:
eslint fixes

Commit:
transport envelope test

base64 committed:
scan and process json files sju crojoisr;

Commit:
cleanup code based on es lint rules

Commit:
cleanup code based on es lint rules - fixes to get test to run

base64 committed:
Merge branch 'pbjx-token' of github.com:gdbots/pbjx-js into pbjx-token

base64 committed:
adjust apiGateway event detection

base64 committed:
add event name assertions when testing determineEventSource

base64 committed:
move aws event constants to seperate file

Commit:
get config test

Commit:
update transport envelope and getconfig

base64 committed:
determineEventSource() test suite should be concrete in its assertions, differentiate api gateway requests from cli lambda invocations

base64 committed:
Merge branch 'pbjx-token' of github.com:gdbots/pbjx-js into pbjx-token

base64 committed:
remove toHttpHeader method from pbjxToken, rename toJson to toJSON, remove public generatePayload method

base64 committed:
remove generateHeader public mthod

base64 committed:
token should be called token

base64 committed:
catch exceptions from jws lib, use toString() method in place of unused getToken

base64 committed:
toJSON should return JWT token in string format

base64 committed:
getPayload can return mixed content type

Commit:
fix transport envelope test

base64 committed:
Merge branch 'pbjx-token' of github.com:gdbots/pbjx-js into pbjx-token

Commit:
simplify toString method

base64 committed:
branch merge

base64 committed:
export PbjxToken from pbjx-js library

Commit:
use constraint for aws-sdk

base64 committed:
detect all internally generated AWS event types

base64 committed:
remove the word event from AWS_EVENT constants as it's implied

base64 committed:
Merge branch 'pbjx-token' of github.com:gdbots/pbjx-js into pbjx-token

base64 committed:
be less speficic in testing for apiGateway Requests as they will reveal more client meta on closer inspection

Commit:
remove duplicate test, move code to correct location

base64 committed:
add event types for aws lex and aws iot-button

base64 committed:
branch merge

gdbrown commented on package.json:
```diff
@@ -18,20 +18,22 @@
     "@gdbots/common": "^0.1.0 || ^0.2.0 || ^0.3.0 || ^0.4.0 || ^1.0.0",
     "@gdbots/pbj": "^0.1.1 || ^0.2.0 || ^0.3.0 || ^0.4.0 || ^1.0.0",
     "@gdbots/schemas": "^1.4.0",
+    "aws-sdk": "^2.6.6",
```
isn't aws on 2.141.0?

gdbrown commented on src/PbjxToken.js:
```diff
@@ -0,0 +1,136 @@
+     * Currently supports: 'exp'
+     * @param {object} tokenData - decoded JWS object
+     */
+    function checkClaims(tokenData) {
```
@acastro move to outside of class... this creates a new function on each token.

gdbrown commented on src/PbjxToken.js:
```diff
@@ -0,0 +1,136 @@
+   * @returns {PbjxToken}
+   */
+  static create(host, content, secret) {
+    function getPayloadHash() {
```
@acastro move to outside of class `const createPayloadHash = (content) => { ...`.  would then not produce a new function/closure on each call.

gdbrown commented on src/PbjxToken.js:
```diff
@@ -0,0 +1,136 @@
+    function getPayloadHash() {
+      return crypto.createHmac('sha256', secret)
+        .update(content)
+        .digest('base64');
```
@acastro you don't need to sign the content, just create the sha256 hash and return the string, no base64 either.

acastro replied:
content is unsigned/salted and now is returned as a hex string

gdbrown commented on src/PbjxToken.js:
```diff
@@ -0,0 +1,136 @@
+
+  /**
+   * Get the decoded header
+   * @returns {string}
```
@acastro this is supposed to be an object correct?

acastro replied:
yes

gdbrown commented on src/PbjxToken.js:
```diff
@@ -0,0 +1,136 @@
+   * Get the decoded payload
+   * @returns {mixed}
+   */
+  getPayload() {
```
@acastro this should always be an object with the host, exp and pbjx correct?

acastro replied:
Yes it was initially intended to return the 'content'  but in this case we'll defer to the JWT payload which is always an object

gdbrown commented on src/aws/lambda/determineEventSource.js:
```diff
@@ -0,0 +1,141 @@
+export const AWS_EVENT = {
```
@acastro change to `export const eventSources = {`

gdbrown commented on src/transports/TransportEnvelope.js:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    const serializer = obj.serializer ? obj.serializer : 'json';
+    const message = TransportEnvelope.getSerializer(serializer).deserialize(obj.message ? obj.message : '');
```
@albertgcsula once you get the json into an object, call fromObject.

gdbrown commented on src/transports/TransportEnvelope.js:
```diff
@@ -0,0 +1,106 @@
+   *
+   * @returns {TransportEnvelope}
+   */
+  static fromObject(obj = {}) {
```
@albertgcsula fromObject needs to address the `is_replay` boolean.  that must be restored otherwise we lose that capability.

gdbrown commented on src/transports/TransportEnvelope.js:
```diff
@@ -0,0 +1,106 @@
+  /**
+   * Returns a json string version of the envelope.
+   *
+   * @return string
```
@albertgcsula wrong doc block format  `@returns {string}`

gdbrown commented on src/transports/TransportEnvelope.js:
```diff
@@ -0,0 +1,106 @@
+  toString() {
+    return JSON.stringify(this);
+  }
+
```
@albertgcsula this class needs a `getMessage()` method to get the message out of the envelope.

gdbrown commented on tests/PbjxToken.test.js:
```diff
@@ -0,0 +1,29 @@
+
+test('JWT HMAC SHA256 Token', (t) => {
+  t.test('Creating JWT token');
+  const host = 'tmzdev.com';
```
@acastro take tmz names out of this repo

gdbrown commented on tests/PbjxToken.test.js:
```diff
@@ -0,0 +1,29 @@
+  const content = ['envelope1', 'envelope2'];
+  const pbjxToken = PbjxToken.create(host, JSON.stringify(content), secret);
+
+  t.test('Verifying JWT token with correct secret');
```
@acastro these tests need to be wrapped into a unit:
```
test('name of test', (t) => {
  // your assertions
  t.end();
});
```

gdbrown commented on tests/aws/lambda/determineEventSource.test.js:
```diff
@@ -0,0 +1,37 @@
+import mockEventSes from './sample-events/ses.json';
+import mockEventSns from './sample-events/sns.json';
+
+function eventTypeMatches(event, expectedEvent) {
```
@acastro simplify to `const eventTypeMatches = (event, expected) => determineEventSource(event) === expected;

gdbrown commented on tests/aws/lambda/determineEventSource.test.js:
```diff
@@ -0,0 +1,37 @@
+  return eventType === expectedEvent;
+}
+
+test('Event Type Detection Tests', async (t) => {
```
@acastro generally prefer to put name of class or function/file being tested with some description... `determineEventSource tests`, etc.

gdbrown commented on tests/aws/lambda/getConfig.test.js:
```diff
@@ -0,0 +1,25 @@
+        {
+          Type: 'String',
+          Name: 'pbjxHost',
+          Value: 'master-api.dev.tmzlabs.com',
```
@albertgcsula no tmz refs here

gdbrown commented on tests/determineEventSource.test.js:
```diff
@@ -0,0 +1,40 @@
+import test from 'tape';
```
@acastro this file is a duplicate?  if not, it should be in same directory as target file being tested.

acastro replied:
moves to aws tests

gdbrown commented on tests/transports/TransportEnvelope.test.js:
```diff
@@ -0,0 +1,32 @@
+
+  t.same(transportEnvelope.isReplay(), false);
+
+  t.end();
```
@albertgcsula need to test that `getMessage()` and `isReplay()` is properly maintained.  those can be separate tests, not all within that same test case.  `TransportEnvelope isReplay tests` for example.

gdbrown reviewed (changes requested):
@acastro and @albertgcsula bunch of notes, mostly small stuff.  please change and ping me when ready.

base64 committed:
jsDoc fixes

base64 committed:
rename AWS_EVENT constant export to eventSources

base64 committed:
library should not refer to the text 'tmz' or any tmz hosts

Commit:
requested updates

Commit:
requested updates - replay

base64 committed:
optimize pbjxToken unit tests

base64 committed:
Merge branch 'pbjx-token' of github.com:gdbots/pbjx-js into pbjx-token

base64 committed:
pr tasks

Commit:
fix from object method

Commit:
cut out from object method

Commit:
if no message or serializer then throw exception

Commit:
argument type for deserialize is object

Commit:
argument type for deserialize is object - rev

gdbrown committed:
aws-mock is indeed not working.  doh!

gdbrown committed:
Merge remote-tracking branch 'origin/pbjx-token' into pbjx-token

gdbrown committed:
aws-mock is indeed not working.  doh!

base64 committed:
alpha sort event source names and registrations

gdbrown committed:
Fix TransportEnvelope now creating with actual deserialized message

also stubbed some fixmes for PbjxToken

base64 committed:
renamve pbjxtokentest for consistency

base64 committed:
PR tasks

base64 committed:
branch merge

base64 committed:
PR tasks

Commit:
aws mock currently not working for ssm

base64 committed:
allow arbitrary exp values, assume all exp values are unix timestamps in seconds

base64 committed:
Merge branch 'pbjx-token' of github.com:gdbots/pbjx-js into pbjx-token

base64 committed:
add support for kid claim

base64 committed:
add support for kid claim to unit tests

base64 committed:
assume 'kid' claim is alaways available, update unit tests to check kid claimwri

acastro commented on package.json:
```diff
@@ -51,6 +54,7 @@
     "eslint-plugin-react": "^7.4.0",
     "faucet": "^0.0.1",
     "isomorphic-fetch": "^2.2.1",
+    "jws": "^3.1.4",
```
Do we need to import JWS twice

base64 committed:
optmize test for amazon SNS event check

gdbrown committed:
added fixme for final tweaks to token and getConfig

Commit:
fix get config expiration time

base64 committed:
add kid to all created tokens, validate kid exists, convienience method for getting kId and exp

base64 committed:
Merge branch 'pbjx-token' of github.com:gdbots/pbjx-js into pbjx-token

base64 committed:
require kid claim

base64 committed:
comment

gdbrown committed:
small code optimizations for tokens, also use standard jwt claims

changing our use of "host" and "pbjx" to use "aud" and "jti" as they are basically what we're using the jwt spec as.  The host aka aud (audience) claim tells us what the intended server endpoint this token is for and is exactly what we want.

The "jti" is a nonce which is also what we are looking for with the pbjx (a hash of the content) which should only be used once.  We combine the aud and iat with that content to further ensure its uniqueness as it's quite possible to have the same payload but not nearly as likely to have the same at exactly the same time for the same aud (endpoint).

By using the reserved jwt claims we might be able to generalize the validation or use other tooling to create/validate the jwt at different layers of the stack.

gdbrown committed:
add PbjxToken creation to HttpTransport when access token is present.

gdbrown committed:
reword comment for eslint rule

gdbrown committed:
0.1.6
